### DISCUSSION TITLE
Receiving a reply

### DISCUSSION THREAD

**Discussion by JaseRandall** [HavenDV/H.Pipes#68] [Q&A] [answered]
Is h.pipes only for one way communication?

I have set up h.pipes to send a message, pipe.SendAsync("Open window") which makes a window open in another process. Easy to setup and works really well, thank you.

However, I can't see any examples or tests that I could use along the lines of:
var dayOfTheWeek = client.SendAsync("What day is it?")

Then in the server code something like:

string QuestionReceived(string msg)
{
  if (msg == "what day is it")
     return "Monday"
}

I have previously done something similar with windows remoting but need to send a msg from a .NET 6 64bit process to a .NET 4.8 32bit process so windows remoting is not possible.

I have seen your IPC project which looks like it allows return values but was hoping stick with h.pipes without need to use a common interface. (Although I guess the interface could simply be one function of string Message(string msg); )

Are there any code examples of how to achieve this kind of two way communication with h.pipes or do I need to use h.IPC?

p.s. I'm not very experienced in this space so I may be way off the mark and misunderstanding what pipes are used for.

**Comment by HavenDV** [comment]
```cs
Server.ClientConnected += async (o, args) =>
        {
            Clients.Add(args.Connection.PipeName);
            UpdateClientList();

            AddLine($"{args.Connection.PipeName} connected!");

            try
            {
                await args.Connection.WriteAsync("Welcome! You are now connected to the server.").ConfigureAwait(false);
            }
            catch (Exception exception)
            {
                OnExceptionOccurred(exception);
            }
        };
```

```cs
await Server.WriteAsync(textBoxMessage.Text).ConfigureAwait(false);
```

https://github.com/HavenDV/H.Pipes/blob/master/src/samples/WindowsFormsApp

**Reply by JaseRandall** [reply]
Thanks for the reply but that's not quite what I was after. I'm wanting the server to return a value in response to a request from the client rather than broadcast a message from the server. (Like what occurs with a HttpClient.GetAsync() request.)

So, something like this:

```
    // on the client side
    var dayOfWeek = await Client.WriteAsync("What day is it?").ConfigureAwait(false);
    Console.WriteLine(dayOfWeek.Result);
```

```
    // on the server side
    private string MessageReceived(string text)
    {
        return "It's Monday.";
    }
```

Or do I need to handle it at the class level in the client. So, something like on the client side:
- have a class level properties `bool MessageRecevied { get; set; }` and string `string Response { get; set; }`
- send the request and loop while `MessageRecevied == false`
- receive the request event back and pass it to the `Response = receivedMessage` property then set `MessageRecevied = true`

Or have I misunderstood the use case for Named Pipes and should be looking for something else?

**Reply by JaseRandall** [reply]
I think I worked out the answer to my question... what I need is synchronous communication rather than asynchronous communication and H.Pipes is only for asynchronous communication?

And after having a look at H.Ipc, it contains an example of returning a person object (and synchronous communication using Named Pipes seems is a lot more complex than both asynchronous communication with Named Pipes and Windows Remoting which I had previously used). So, should use H.Ipc instead?

**Reply by JaseRandall** [reply, marked as answer]
I got my project working with [https://github.com/HavenDV/H.Ipc](https://github.com/HavenDV/H.Ipc).

Works really well. Thank you Konstantin!!
